### PR TITLE
BlockList: Ensure element styles (and svg) are always appended at the end of the document

### LINKED ISSUE #36053: Container Style Element Breaks Expected CSS Selectors
### Description

I've started trying to use `theme.json` which triggers new layout tools for Group blocks. Among these are an inlined `style` element immediately following the Group block:

```
<div class="wp-container-617ac576851fb wp-block-group">...</div>
<style>.wp-container-617ac576851fb {} ...</style>
```

This breaks what has become a really helpful theming technique for CSS Selectors:

```
/* global vertical spacing */
.entry-content > * + * {
    margin-top: 1.5em;
}
.alignfull + alignfull {
    margin-top: 0;
}
```

This worked with the old markup that would be this:

```
<div class="wp-block-group alignfull"></div>
<div class="wp-block-group alignfull"></div>
```

It no longer works because the markup is now:

```
<div class="wp-block-group alignfull"></div>
<style></style>
<div class="wp-block-group alignfull"></div>
```

### Step-by-step reproduction instructions

1. Add a style to the site that selects the adjacent sibiling of a group block. e.g. `.wp-block-group + * { background: red !important; }`
2. Create a page with two Group blocks.
3. The second group block should have a red background.
4. Add theme.json to theme and resave post.
5. The second group block no longer will have a red background.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

WordPress 5.8.1, no Gutenberg plugin, traditional theme with theme.json file

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Fixes #36053

Update the `BlockList`'s `Root` component to always output element styles (e.g. `style` and `svg` elements for layout, element, and duotone block supports) to the end of the document. To do so, add an extra `div` to the end of `Root` to ensure that any changes to the `useInnerBlocksProps` element occur prior to the injected `style` and `svg` tags.

## Why?

The reason this change is needed is fairly subtle. React's `createPortal` appends nodes as the list child of an element. In an already loaded post or page, the inner blocks for the block list will all be rendered at the top, and the injected `style` tags will appear at the bottom of the DOM.

Something interesting happens when we go to make changes, though. When hovering over the end of the document and going to insert blocks, the block appender will be rendered _after_ the injects `style` tags in the DOM. Then, when we go to insert blocks, inserted blocks will _also_ be rendered after the injected styles. Whereas, what we expect to happen is that the styles would still appear at the end.

The reason why, I think, is because after the original injection, when React rerenders elements into the portal, it doesn't attempt to remount into the portal, rather it makes updates in place since the elements have already been rendered into the DOM. This means that subsequent additions are added to the end, but we wind up in an unexpected situation where `style` tags wind up "randomly" sitting between other blocks once we go to make changes to blocks within a document.

To avoid this collision between an element where lots of mutation is happening due to addition, removal, or moving of blocks, and the injection of styles and svg tags, a solution is to create a separate `div` as the entrypoint for portal-based injection of style and svg tags, and have this always sit at the bottom of the document.

For further explanation as to why the style and svg tags sitting between blocks is an issue, see the discussion in: https://github.com/WordPress/gutenberg/issues/36053 — the TL;DR is that the presence of these tags can break CSS selectors as such sibling, first / last child, etc.

## How?

* Add an extra `div` to the bottom of the `Root`.
* Move the `setElement` ref state function to this node.

## Testing Instructions

1. Add a variety of Group and Row blocks to a post, page, or template and adjust layout and element settings (like color, custom constrained width, content justification), and ensure the styles all render correctly. Add an Image block and set custom Duotone on the image, also check that this renders correctly.
2. Save and reload the page, ensure rendering is still correct.
3. Inspect the document and observe that there should be `style` tags in a `div` at the end of the document (beneath the block content).
4. Go to insert blocks at the end of the document, using the block inserter button within the editor canvas. Observe that the newly inserted blocks should still be inserted _before_ the injected `style` tags that contain the `wp-container-123` etc rules.
5. Test in a range of browsers to make sure this doesn't introduce any regressions (of note, I noticed that we can't add the `hidden` attribute to the container `div`, because Firefox requires the container to be visible in order for the `svg` elements to be accessible for rendering Duotone). As far as I can tell, the current state of this PR doesn't cause any styling issues, but it would be good to confirm via manual testing.

<details>

<summary>
Here is some test markup to begin with (you'll need to add your own Image blocks with Duotone for testing):
</summary>

```html

<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap","justifyContent":"center"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>A paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Another paragraph in a row</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"elements":{"link":{"color":{"text":"#ff4545"}}}},"layout":{"type":"constrained","contentSize":"320px","wideSize":"480px"}} -->
<div class="wp-block-group has-link-color"><!-- wp:paragraph -->
<p>A paragraph in a constrained Group block</p>
<!-- /wp:paragraph -->

<!-- wp:cover {"dimRatio":50,"overlayColor":"luminous-vivid-orange","minHeight":123,"minHeightUnit":"px","isDark":false} -->
<div class="wp-block-cover is-light" style="min-height:123px"><span aria-hidden="true" class="wp-block-cover__background has-luminous-vivid-orange-background-color has-background-dim"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph -->
<p>A cover in a <a href="http://wordpress.org" data-type="link" data-id="wordpress.org">constrained</a> Group block</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover -->

<!-- wp:cover {"dimRatio":50,"overlayColor":"luminous-vivid-orange","minHeight":123,"minHeightUnit":"px","isDark":false,"align":"wide"} -->
<div class="wp-block-cover alignwide is-light" style="min-height:123px"><span aria-hidden="true" class="wp-block-cover__background has-luminous-vivid-orange-background-color has-background-dim"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph -->
<p>A cover in a constrained Group block</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover --></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"position":{"type":"sticky","top":"0px"},"spacing":{"padding":{"top":"var:preset|spacing|30","bottom":"var:preset|spacing|30","left":"var:preset|spacing|30","right":"var:preset|spacing|30"}}},"backgroundColor":"vivid-red","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-vivid-red-background-color has-background" style="padding-top:var(--wp--preset--spacing--30);padding-right:var(--wp--preset--spacing--30);padding-bottom:var(--wp--preset--spacing--30);padding-left:var(--wp--preset--spacing--30)"><!-- wp:paragraph -->
<p>A paragraph in a sticky group</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

```

</details>

## Screenshots or screencast 

The below screenshots show the in editor markup for a post just after a paragraph block has been inserted at the end of the document.

Note that in the "before" screenshot, there is a Figure element for the image block, then a series of `style` and `svg` tags, then the paragraph block, then the block appender.

In the "after" screenshot, in the top area there is the Figure element for thee image block, then the paragraph block, then the block appender. And below all of them is a separate `div` to contain all the `style` and `svg` tags.

| Before | After |
| --- | --- |
| image | image |

### CODE REVIEW

**Review by youknowriad — COMMENTED** [review]
This looks good to me but I'd like for @ellatrix to confirm here. 

The other thing here is that, there's no a small chance that this extra div impacts CSS that is injected by the theme, things like last-child selectors... Maybe the percentage of themes this impacts is small?

**Comment by andrewserong** [comment]
Thanks for taking a look @youknowriad! 

> The other thing here is that, there's no[w] a small chance that this extra div impacts CSS that is injected by the theme, things like last-child selectors... Maybe the percentage of themes this impacts is small?

True — I think the likelihood is probably fairly minimal, and on balance I think the behaviour with this PR is probably preferable and keeps the injected styles and svgs in a consistent place at least. My main thinking here is that this extra div is now a sibling of `is-root-container`, and in the post editor we already have quite a few other siblings of that container preceding it. The following screengrab highlights the new `div`, with the `is-root-container` preceding it, along with all the other stuff above (direct children of `body`):

image

> This looks good to me but I'd like for @ellatrix to confirm here.

Agreed, it'd be good to get another pair of eyes on this if we can. Since this PR is a change that resolves a fairly subtle issue, and in turn could cause other fairly subtle issues, to be on the safe side I'll just give a wider ping to @WordPress/gutenberg-core in case there's anyone who has further thoughts about the markup here 🙂

**Comment by andrewserong** [comment]
Just gave this a fresh rebase to fix a conflict after #53875 landed, so this should be good to review again now.

**Comment by Mamaduka** [comment]
There's also the #52888, which proposes a new system for outputting styles and SVGs.

**Review by youknowriad — APPROVED** [review]
Approving this cause overall it's seems like an improvement for the current way we do things. We can look at the further improvements separately.

**Comment by ellatrix** [comment]
Sure this is fine, but outputting extra divs. You all know how much I hate extra divs 😅
Would love if someone could have a look at my proposal #52888. We can get rid of `BlockList.__unstableElementContext` entirely.

**Comment by andrewserong** [comment]
Thanks for the reviews and feedback, folks! I'll merge this in now since it fixes the immediate issue for the moment. 

> Would love if someone could have a look at my proposal https://github.com/WordPress/gutenberg/pull/52888. We can get rid of BlockList.__unstableElementContext entirely.

Agreed, let's take a closer look at #52888 for a better long-term solution 👍

**Pull request merged**
